src/MongoDB.Driver.Core/Core/Clusters/Cluster.cs

```diff
@@ -507,25 +507,21 @@ public void WaitingForDescriptionToChange()
             private IServerSelector DecorateSelector(IServerSelector selector)
             {
                 var settings = _cluster.Settings;
-                if (settings.PreServerSelector != null || settings.PostServerSelector != null)
+                var allSelectors = new List<IServerSelector>();
+                if (settings.PreServerSelector != null)
                 {
-                    var allSelectors = new List<IServerSelector>();
-                    if (settings.PreServerSelector != null)
-                    {
-                        allSelectors.Add(settings.PreServerSelector);
-                    }
+                    allSelectors.Add(settings.PreServerSelector);
+                }
 
-                    allSelectors.Add(selector);
+                allSelectors.Add(selector);
 
-                    if (settings.PostServerSelector != null)
-                    {
-                        allSelectors.Add(settings.PostServerSelector);
-                    }
-
-                    return new CompositeServerSelector(allSelectors);
+                if (settings.PostServerSelector != null)
+                {
+                    allSelectors.Add(settings.PostServerSelector);
                 }
 
-                return selector;
+                allSelectors.Add(new LatencyLimitingServerSelector(settings.LocalThreshold));
```

Comment:
Additional validation on presentence `LatencyLimitingServerSelector` in the last selector can be added, however, I think the current implementation is also goodlooking